odoo/modules/migration.py

```diff
@@ -79,6 +79,59 @@ def __init__(self, cr, graph):
         self.graph = graph
         self.migrations = defaultdict(dict)
         self._get_files()
+        self._init_migration_history()
+
+    def _init_migration_history(self):
+        """
+        Create stages for a module graph to allow restarting the migration.
+
+        For each module, an ir_logging entry is created for the three possible stages of the migration
+        (pre, post, end) and the `line` column will contain either an empty string (the scripts have
+        not yet been applied for that stage) or a non-empty string (the scripts have been applied).
+        This trick is necessary for 'end' scripts, since we need to possibly restart them even if
+        the module's version is already up-to-date (since the `latest_version` is updated after
+        post scripts are run on an ir_module record).
+
+        Note that this mechanism assumes previous migrations were clean - there should be no
+        ir_logging entries of level `UPGRADE` prior to migration.
+
+        Also, explicit commits in a migration script will very, very probably break this feature
+        and prevent restarting a migration properly.
+        """
+        if not tools.table_exists(self.cr, 'ir_logging'):
+            # database is brand new, skip
+            return
+        for pkg in self.graph:
+            if parse_version(pkg.installed_version) > parse_version(release.major_version):
```

Comment:
should probably check version currently in db for that module in case we ever decide to bump a manifest file